client/xdag_config.c

```diff
@@ -0,0 +1,535 @@
+	node->keys[node->number_keys-1] = key;
+	node->key_names = (char **)realloc(node->key_names, node->number_keys * sizeof(char *));
+	node->key_names[node->number_keys-1] = strdup(key->name);
+}
+
+static KEY *xdag_config_read_key(FILE *fp)
+{
+    long offset = ftell(fp);
+    CFG_CONTENT_TYPE type;
+    char *line = NULL;
+
+    while((line = trim(readline(fp))) != NULL)
+    {
+        type = xdag_get_content_type(line);
+        if(type == TYPE_KEY)
+        {
+            KEY *key;
+            char *key_name;
+            char *p = strchr(line, '=');
+            *p = '\0';
+            key_name = trim(line);
+            key = xdag_add_key(key_name);
+            key->value = trim(strdup(p+1));
+            free(line);
+            return key;
+        }
+        else if(type == TYPE_NODE)
+        {
+            free(line);
+            fseek(fp, offset, SEEK_SET);
+            return NULL;
+        }
+        else
+        {
+            free(line);
+            offset = ftell(fp);
+            continue;
+        }
+    }
+
+    return NULL;
+}
+
+
+static NODE *xdag_config_read_node(FILE *fp)
+{
+	NODE *node = NULL;
+    KEY *key = NULL, *existKey;
+    char *node_name = xdag_config_read_node_name(fp);
+    if(node_name != NULL)
+    {
+    	node = xdag_add_node(node_name);
+        free(node_name);
+
+        while((key = xdag_config_read_key(fp)) != NULL)
+        {
+            existKey = xdag_config_get_key(node, key->name);
+            if(existKey != NULL)
+            {
+            	xdag_config_free_key(key);
+            }
+            else
+            {
+            	xdag_config_add_key(node,key);
+            }
+        }
+        return node;
+    }
+    return NULL;
+}
+
+
+static void xdag_config_read(CFG *config, FILE *fp)
+{
+    NODE *node = NULL, *existNode;
+
+    while((node = xdag_config_read_node(fp)) != NULL)
+    {
+    	existNode = xdag_get_node(config, node->name);
+        if(existNode != NULL)
+        {
+            xdag_config_free_node(node);
+        }
+        else
+        {
+        	xdag_config_add_node(config, node);
+        }
+    }
+}
+
+static CFG *xdag_config_init(const char *path)
+{
+    CFG *cfg = (CFG *)malloc(sizeof(CFG));
+
+    if(path != NULL)
+    {
+    	cfg->path = strdup(path);
+    }
+    else
+    {
+    	cfg->path = NULL;
+    }
+    cfg->number_nodes = 0;
+    cfg->nodes = NULL;
+    cfg->node_names = NULL;
+    return cfg;
+}
+
+
+static void* xdag_config_open(const char *path)
+{
+    CFG *cfg = xdag_config_init(path);
+    FILE *fp;
+
+    if(cfg->path != NULL)
+    {
+        fp = fopen(cfg->path, "r");
+        if(fp != NULL)
+        {
+        	xdag_config_read(cfg, fp);
+            fclose(fp);
+        }
+    }
+    return (void*)cfg;
+}
+
+
+static void xdag_config_close(void *cfg)
+{
+    CFG *config = (CFG *)cfg;
+    xdag_config_free(config);
+}
+
+
+
+char * get_pool_config(const char *path){
+	static char result[80];
+	if(path) {
```

Comment:
why don't use strcat?
and i see that you doesn't check the buffer overflow!

Reply:
I rewrote this function.